e2e/test/iothub/SasCredentialAuthenticationTests.cs

```diff
@@ -0,0 +1,225 @@
+                Configuration.IoTHub.GetIotHubHostName(),
+                new AzureSasCredential(signature));
+
+            // act
+            HttpOperationResponse<ThermostatTwin, DigitalTwinGetHeaders> response = await digitalTwinClient
+                .GetDigitalTwinAsync<ThermostatTwin>(testDevice.Id)
+                .ConfigureAwait(false);
+            ThermostatTwin twin = response.Body;
+
+            // assert
+            twin.Metadata.ModelId.Should().Be(thermostatModelId);
+
+            // cleanup
+            await testDevice.RemoveDeviceAsync().ConfigureAwait(false);
+        }
+
+        [Ignore]
+        [LoggedTestMethod]
+        public async Task Service_Amqp_SasCredentialAuth_Success()
+        {
+            // arrange
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, _devicePrefix).ConfigureAwait(false);
+            using DeviceClient deviceClient = testDevice.CreateDeviceClient(Client.TransportType.Mqtt);
+            await deviceClient.OpenAsync().ConfigureAwait(false);
+
+            string signature = Configuration.IoTHub.GetIotHubSharedAccessSignature(TimeSpan.FromHours(1));
+            using var serviceClient = ServiceClient.Create(
+                Configuration.IoTHub.GetIotHubHostName(),
+                new AzureSasCredential(signature),
+                TransportType.Amqp);
+
+            // act
+            await serviceClient.OpenAsync().ConfigureAwait(false);
+            using var message = new Message(Encoding.ASCII.GetBytes("Hello, Cloud!"));
+
+            await serviceClient.SendAsync(testDevice.Id, message);
+
+            // cleanup
+            await testDevice.RemoveDeviceAsync().ConfigureAwait(false);
+        }
+
+        [Ignore]
+        [LoggedTestMethod]
+        public async Task Service_Amqp_SasCredentialAuth_Renewed_Success()
+        {
+            // arrange
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, _devicePrefix).ConfigureAwait(false);
+            using DeviceClient deviceClient = testDevice.CreateDeviceClient(Client.TransportType.Mqtt);
+            await deviceClient.OpenAsync().ConfigureAwait(false);
+
+            string signature = Configuration.IoTHub.GetIotHubSharedAccessSignature(TimeSpan.FromHours(-1));
+            var sasCredential = new AzureSasCredential(signature);
+            using var serviceClient = ServiceClient.Create(
+                Configuration.IoTHub.GetIotHubHostName(),
+                sasCredential,
+                TransportType.Amqp);
+
+            // act
+            await serviceClient.OpenAsync().ConfigureAwait(false);
+            using var message = new Message(Encoding.ASCII.GetBytes("Hello, Cloud!"));
+
+            try
+            {
+                await serviceClient.SendAsync(testDevice.Id, message);
+            }
+            catch (Exception)
+            {
+                // TODO: Test and catch the correct exception.
```

Comment:
Which exception is expected here?

Reply:
I will know on testing. But expecting IotHubUnauthorized